File: 2001-3000/2517-maximum-tastiness-of-candy-basket/2517-maximum-tastiness-of-candy-basket.java
class Solution {
    public int maximumTastiness(int[] price, int k) {
        Arrays.sort(price);
        int n = price.length;
        int left = 0, right = (price[n - 1] - price[0]) / (k - 1);
        while(left < right) {
            int mid = (right + left + 1) >> 1;
            if(check(price, mid, k)) left = mid;
            else right = mid - 1;
        }
        return right;
    }
    private boolean check(int[] price, int target, int k) {
        int prev = 0;
        for(int x : price) {
            if(x < prev) continue;
            if(--k == 0) return true;
            prev = x + target;
        }
        return false;
    }
}

File: 2001-3000/2517-maximum-tastiness-of-candy-basket/2517-maximum-tastiness-of-candy-basket.cpp
class Solution {
public:
bool check(vector<int>& arr, int k,int m){
    int n = arr.size();
    int curr=arr[0];
    k--;
    int r=1;
    while(r<n){
        if(arr[r]-curr>=m){
            k--;
            curr=arr[r];
        }
        r++;
    }
    if(k<=0) return true;
    return false;
}

    int maximumTastiness(vector<int>& arr, int k) {
        int n = arr.size();
        sort(arr.begin(),arr.end());
        int l=0;
        int h=arr[n-1]-arr[0];
        while(l<=h){
            int m =(l+h)/2;
            if(check(arr,k,m)){
                l=m+1;
            }
            else h= m-1;
        }
        return h;
    }
};